internal/resources/metal/virtual_circuit/resource.go

```diff
@@ -0,0 +1,459 @@
+			vncr.VlanVirtualCircuitCreateInput.NniVlan = metalv1.PtrInt32(int32(nniVlan.(int)))
+		} else {
+			vncr.VrfVirtualCircuitCreateInput.NniVlan = int32(nniVlan.(int))
+		}
+	}
+
+	conn, _, err := client.InterconnectionsApi.GetInterconnection(ctx, connId).Execute()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if conn.GetStatus() == string(metalv1.VLANVIRTUALCIRCUITSTATUS_PENDING) {
+		return diag.Errorf("Connection request with name %s and ID %s wasn't approved yet", conn.GetName(), conn.GetId())
+	}
+
+	vc, _, err := client.InterconnectionsApi.CreateInterconnectionPortVirtualCircuit(ctx, connId, portId).VirtualCircuitCreateInput(vncr).Execute()
+	if err != nil {
+		log.Printf("[DEBUG] Error creating virtual circuit: %s", err)
+		return diag.FromErr(err)
+	}
+
+	var vcId string
+
+	if vc.VlanVirtualCircuit != nil {
+		vcId = vc.VlanVirtualCircuit.GetId()
+	} else {
+		vcId = vc.VrfVirtualCircuit.GetId()
+	}
+
+	// TODO: offer to wait while VCStatusPending
+	createWaiter := getVCStateWaiter(
+		ctx,
+		client,
+		vcId,
+		d.Timeout(schema.TimeoutCreate)-30*time.Second,
+		[]string{string(metalv1.VLANVIRTUALCIRCUITSTATUS_ACTIVATING)},
+		[]string{string(metalv1.VLANVIRTUALCIRCUITSTATUS_ACTIVE)},
+	)
+
+	_, err = createWaiter.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("Error waiting for virtual circuit %s to be created: %s", vcId, err.Error())
+	}
+
+	d.SetId(vcId)
+
+	return resourceMetalVirtualCircuitRead(ctx, d, meta)
+}
+
+func resourceMetalVirtualCircuitRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*config.Config).NewMetalClientForSDK(d)
+	vcId := d.Id()
+
+	vc, _, err := client.InterconnectionsApi.GetVirtualCircuit(ctx, vcId).
+		Include([]string{"project", "virtual_network", "vrf"}).
+		Execute()
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// TODO: use API field from VC responses when available The regexp is
+	// optimistic, not guaranteed. This affects resource imports. "port" is not
+	// in the Includes above to assure the Href needed below.
+	var portHref string
+
+	if vc.VlanVirtualCircuit != nil {
+		portHref = vc.VlanVirtualCircuit.Port.GetHref()
+	} else {
+		portHref = vc.VrfVirtualCircuit.Port.GetHref()
+	}
+	connectionID := "" // vc.Connection.ID is not available yet
+	portID := ""       // vc.Port.ID would be available with ?include=port
+	connectionRe := regexp.MustCompile("/connections/([0-9a-z-]+)/ports/([0-9a-z-]+)")
+	matches := connectionRe.FindStringSubmatch(portHref)
+	if len(matches) == 3 {
+		connectionID = matches[1]
+		portID = matches[2]
+	} else {
+		log.Printf("[DEBUG] Could not parse connection and port ID from port href %s", portHref)
+	}
+	errs := &multierror.Error{}
+
+	if connectionID != "" {
+		multierror.Append(errs, d.Set("connection_id", connectionID))
+	}
+	d.Set("port_id", portID)
+
+	if vc.VlanVirtualCircuit != nil {
+		multierror.Append(errs, d.Set("project_id", vc.VlanVirtualCircuit.Project.GetId()))
```

Comment:
multierror.Append is variadic: https://github.com/hashicorp/go-multierror/blob/main/append.go#L14

This could be a comma list of d.Sets.

I don't know if d.Set will include the fieldname in the error messages, like (I think) SetMap does.
We may have unnamed error messages that are hard to pin down in this list.

Reply:
SetMap does not do anything special to the error it gets from `d.Set`.  It just passes it to `multierror.Append` as was done here.  I've since switched to `append` and `errors.Join` from the standard library.  That is also variadic but IMO the repetition is worthwhile here and reduces the chance that someone copy-pastes a line to add support for a new attribute and happens to paste it outside of the `append()` call.